pkg/util/atomic_copier.go

```diff
@@ -0,0 +1,132 @@
+package util
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	"k8s.io/klog/v2"
+)
+
+var (
+	cpArgs = []string{
+		"--verbose",
+		"--recursive",
+		"--preserve",
+		"--reflink=auto",
+	}
+)
+
+// AtomicCopier performs a two operation: copies source path to an intermediate location,
+// then renames from intermediate to final destination.
+// On Unix systems, the rename operation is atomic. This ensures that the file
+// is either fully renamed or not at all, which helps prevent issues like partial
+// file copies in cases of power failure or unexpected program termination.
+type AtomicCopier struct {
+	Source      string
+	Destination string
+
+	intermediatePath string
+}
+
+func (c *AtomicCopier) CopyToIntermediate() error {
+	if c == nil {
+		return nil
+	}
+	c.intermediatePath = fmt.Sprintf("%s.tmp", c.Destination)
+	if exists, err := PathExists(c.intermediatePath); err != nil {
+		return err
+	} else if exists {
+		if err := os.RemoveAll(c.intermediatePath); err != nil {
+			klog.Errorf("Failed to remove intermediate path %q which already existed: %v", c.intermediatePath, err)
+			return fmt.Errorf("failed to remove %q: %w", c.intermediatePath, err)
+		}
+	}
```

Comment:
I'm not living in peace with this approach of adding `.tmp` suffix. It seems to be too common for the generic `util` code. 
Can we do something like `mktemp` and return a fatal error if the destination exists? I'm a bit wary of deleting files while copying them.